contents/java/Java.md

```diff
@@ -14,6 +14,275 @@ Authors: John Doe, Jane Doe
+optimally be solved by modifying the field visibility to `protected` or `public`, sometimes it is not possible to do so
+becuase you might not have any access to the code (for example codes in libraries or frameworks).
+
+For the sake of simplicity, let use the previous example of a `Animal`. The class definition can look like this.
+
+```java
+
+public class Animal {
+
+    private int age;
+    private boolean isAlive;
+    public final String name;
+
+    public Animal(String nameOfAnimal) {
+        name = nameOfAnimal;
+        age = 0;
+        isAlive = true;
+    }
+
+    public void onUpdate() {
+        age++;
+
+        if (age > 100) {
+            setDead();
+        }
+    }
+
+    private void setDead() {
+        isAlive = false;
+    }
+}
+```
+
+And let us assume that you for some reason cannot modify this code. But you are interested in making a new class `Sheep`
+that extends `Animal` and do something when his age reaches certain threshold. But the annoying thing is that somebody
+decided that it is a good idea to make the age value `private` instead of `protected` in a top-level class such as this!
+So you cannot access the age of your `Sheep` even though it is an `Animal`.
+
+This of course can be solved by reflections as follows:
+
+```java
+
+public class Sheep extends Animal {
+    private boolean hasWool;
+
+    public Sheep(String animalName) {
+        super(animalName);
+        hasWool = false;
+    }
+
+    @Override
+    public void onUpdate() {
+        super.onUpdate();
+
+        int age = getPrivateAge();
+
+        if (age > 20) {
+            setHasWool();
+        }
+    }
+
+    private void setHasWool() {
+        hasWool = true;
+    }
+
+    private int getPrivateAge() {
+        // Since class is a reserved keyword, we use clazz in variable names
+        Class<Animal> animalClazz = Animal.class;
+
+        try {
+            Field f = animalClazz.getDeclaredField("age");
+            f.setAccessible(true);
+            return (int) f.get(this);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        return 0;
+    }
+}
+```
+
+### Disadvantages of reflections
+
+While Java reflections is powerful, you should not immediately jump on the reflections ship. This is because there are
+some drawbacks whenever you use reflections in your project. The following are some points you should consider before
+using Java reflections:
+
+* Reflections converts a compile-time error to a potentially destructive run-time error.
+
+Compile time errors are easy to catch. Whenever you compile your code, the compiler cleverly spots any error you
+missed and points it out (along with line number and other useful information) to you before quitting. But by using
+reflections, you are bypassing these checks because there is no way to check such errors during compile time. Runtime
```

Comment:
"such errors" sounds like you are referring to compile time errors rather than errors in the code. Maybe add a sentence like "These uncaught errors may cause your program to fail during runtime instead, turning into runtime errors." so that the paragraph flows better.